### PR TITLE
Part 14. BidderRequestHook implementation for OpenRTB2 Blocking module

### PR DESCRIPTION
**Pull request by 4lexvav**
Module PRD: https://docs.google.com/document/d/1TpB_I-Oft0LpZuKvVsaLAozOe1u52tpPGkB_jZfnMXo/edit?usp=sharing

### CODE REVIEW

**Comment by 4lexvav** [comment]
[Implement BidRequestHook for ortb2blocking module](https://trello.com/c/sDQYxKl6/25-implement-bidrequesthook-for-ortb2blocking-module)

**Comment by VeronikaSolovei9** [comment]
Are there any requirements or documentation related to changes in this PR? It would help a lot to review it and understand how it works.
Also question: Part 15 PR https://github.com/pi-prebid/prebid-server/pull/24 is an integration of changes from this PR into PBS code, correct? And also changes from PRs Part 9 and Part 13? Any new code there?

**Comment by 4lexvav** [comment]
@VeronikaSolovei9 I added a link to module PRD in the description of this PR.
Yes, you are right. At the moment Part 15 PR #24 requires the `bidder-response` hook and the current branch as the dependencies, but once we merge all modularity PRs into the master branch, we can simplify it.

**Comment by legendko** [comment]
> Are there any requirements or documentation related to changes in this PR? It would help a lot to review it and understand how it works. Also question: Part 15 PR #24 is an integration of changes from this PR into PBS code, correct? And also changes from PRs Part 9 and Part 13? Any new code there?

I'll add that there is plenty of new code that needs to be reviewed in Part 15 PR :grin: 
that pr has the concrete implementation of **raw bidder response hook** for this module
while the current pr has the concrete implementation of **bidder request hook** for this module

File: modules/prebid/ortb2blocking/module_test.go
Comment (VeronikaSolovei9) [inline modules/prebid/ortb2blocking/module_test.go]: This line throws compilation error. Works with this change: `result, err := Builder(nil, moduledeps.ModuleDeps{})`

**Comment by VeronikaSolovei9** [comment]
Thank you for the link to the documentation. It makes more sense now. And thank you for the comprehensive unit tests, it also helps to understand how it works.

**Comment by VeronikaSolovei9** [comment]
Just for my knowledge, can you please help to understand where is the functionality of unwanted bid removal, if it's in this PR.  F.i this requirement states 
```
As a publisher, I need Prebid Server to enforce that bids in certain blocked advertiser categories are removed from the auction for particular bidders so I can prevent unwanted ads from appearing on my site.

```
It's probably in the Part 15.

**Comment by legendko** [comment]
> Just for my knowledge, can you please help to understand where is the functionality of unwanted bid removal, if it's in this PR. F.i this requirement states
> 
> ```
> As a publisher, I need Prebid Server to enforce that bids in certain blocked advertiser categories are removed from the auction for particular bidders so I can prevent unwanted ads from appearing on my site.
> ```
> 
> It's probably in the Part 15.

@VeronikaSolovei9 correct, thats on Part 15.

File: modules/prebid/ortb2blocking/module_test.go
Comment (4lexvav) [reply modules/prebid/ortb2blocking/module_test.go]: right, thanks for noticing, forgot to change this after merging changes from the base branch

File: modules/prebid/ortb2blocking/module_test.go
Comment (VeronikaSolovei9) [inline modules/prebid/ortb2blocking/module_test.go:363]: Can you please add a test case to explicitly test default config (not bidder-specific), something like this: 
```
{
			description: "Payload changed after successful BidderRequest hook execution for default config",
			bidder:      bidder,
			config:      json.RawMessage(`{"attributes": {"badv": {"enforce_blocks": true, "block_unknown_adomain": true, "blocked_adomain": ["a.com","b.com","c.com"]}}}`),
			bidRequest: &openrtb2.BidRequest{
				Imp: []openrtb2.Imp{
					{
						ID:    "ImpID1",
						Audio: &openrtb2.Audio{},
					},
				},
			},
			expectedBidRequest: &openrtb2.BidRequest{
				BAdv: []string{bAdvA, bAdvB, bAdvC},
				Imp: []openrtb2.Imp{
					{
						ID:    "ImpID1",
						Audio: &openrtb2.Audio{},
					},
				},
			},
			expectedHookResult: hookstage.HookResult[hookstage.BidderRequestPayload]{
				ModuleContext: map[string]interface{}{
					bidder: blockingAttributes{
						badv:  []string{bAdvA, bAdvB, bAdvC},
						btype: map[string][]int{},
						battr: map[string][]int{},
					},
				},
			},
			expectedError: nil,
		},
```

**Review by VeronikaSolovei9 — COMMENTED** [review]
Added a comment about unit test adding. Other than that code looks good.

File: modules/prebid/ortb2blocking/module_test.go
Comment (4lexvav) [reply modules/prebid/ortb2blocking/module_test.go:363]: Sure, I added it, thanks for providing the test case.

File: modules/prebid/ortb2blocking/module.go
Comment (mansinahar) [inline modules/prebid/ortb2blocking/module.go]: Can you please rename these in camelCase style to be consistent with the field in the `BidRequest` struct? So like: `bAdv`, `bApp`, `bCat`, etc

File: modules/prebid/ortb2blocking/hook_bidderrequest.go
Comment (mansinahar) [inline modules/prebid/ortb2blocking/hook_bidderrequest.go:360]: Is it possible for `m` to have any media types that aren't one of type audio, banner, native or video? If not, can't we just go through the map and concatenate all the strings without first checking if they are of one of those 4 media type strings?

**Comment by VeronikaSolovei9** [comment]
Just realized, question: if a blocking attribute is specified in incoming request - it will be replaced with the data from config. Is it a correct behavior? Should it be merged with a config data? Cannot find related requirement in the doc.

File: modules/prebid/ortb2blocking/hook_bidderrequest.go
Comment (mansinahar) [inline modules/prebid/ortb2blocking/hook_bidderrequest.go]: First of all, I really like how clean this method is. Thanks :) Can we please also move the logic related to category taxonomy in a separate function so that this function is just calling update function for all these diff fields and doesn't have specific logic to any of them here?

Comment (4lexvav) [reply modules/prebid/ortb2blocking/hook_bidderrequest.go]: Of course, updated ✔️

File: modules/prebid/ortb2blocking/module.go
Comment (4lexvav) [reply modules/prebid/ortb2blocking/module.go]: Renamed ✔️

File: modules/prebid/ortb2blocking/hook_bidderrequest.go
Comment (4lexvav) [reply modules/prebid/ortb2blocking/hook_bidderrequest.go:360]: I see your point. Actually, in the first implementation, I did exactly as you suggest, however later I realized, that the map does not preserve the sort order of the elements, which is required by tests where I compare hook execution results: https://github.com/pi-prebid/prebid-server/blob/de278624a7e7774d1b89518d0543752210826e55/modules/prebid/ortb2blocking/module_test.go#L357
Thus without this trick, I get failed tests.

**Comment by 4lexvav** [comment]
@VeronikaSolovei9 you are right, I found the next requirement in the doc:
> Parameters already present on the incoming request take precedence over parameters generated by the module.

Also, I found that PBS-Java has recently been updated accordingly: https://github.com/prebid/prebid-server-java/pull/2006
So I made appropriate changes.

File: modules/prebid/ortb2blocking/hook_bidderrequest.go
Comment (VeronikaSolovei9) [inline modules/prebid/ortb2blocking/hook_bidderrequest.go:316]: Cool!

**Comment by VeronikaSolovei9** [comment]
Thank you for the changes! Ready to approve it after Mansi confirms all her comments are properly addressed.

File: modules/prebid/ortb2blocking/hook_bidderrequest.go
Comment (mansinahar) [reply modules/prebid/ortb2blocking/hook_bidderrequest.go:360]: Ah ok! Thanks for explaining @4lexvav. That makes sense

**Review by mansinahar — APPROVED** [review]

**Review by VeronikaSolovei9 — APPROVED** [review]

**Comment by 4lexvav** [comment]
PR created in the upstream repo

**Pull request closed without merging**
